wallet/test/Test/Pos/Wallet/Web/Methods/PaymentSpec.hs

```diff
@@ -70,22 +92,29 @@ oneNewPaymentSpec = walletPropertySpec oneNewPaymentDesc $ do
         Left er -> stopProperty $ "Couldn't compute tx fee by tx, reason: " <> pretty er
         Right x -> pure x
     let outAddresses = map (fst . view _TxOut) $ toList $ _txOutputs $ taTx txAux
-    changeAddr <- expectedOne "expected one change address" (filter (/= dstAddr) outAddresses)
+    let changeAddrs = outAddresses \\ dstAddrs
+    assertProperty (length changeAddrs <= 1) $
+        "Expected at most one change address"
 
     -- Validate balances
-    expectedAddrBalance dstAddr coins
-    expectedAddrBalance srcAddr (mkCoin 0)
-    changeBalance <- getBalance changeAddr
+    mapM_ (uncurry expectedAddrBalance) $ zip dstAddrs coins
+    when (policy == OptimizeForSecurity) $
+        expectedAddrBalance srcAddr (mkCoin 0)
+    changeBalance <- mkCoin . fromIntegral . sumCoins <$> mapM getBalance changeAddrs
     assertProperty (changeBalance <= initBalance `unsafeSubCoin` fee) $
         "Minimal tx fee isn't satisfied"
 
     -- Validate that tx meta was added when transaction was processed
-    when (walId /= dstWalId) $ do
-        txMetasSource <- maybeStopProperty "Source wallet doesn't exist" =<< lift (WS.getWalletTxHistory walId)
-        void $ expectedOne "TxMeta for source wallet" txMetasSource
 
-        txMetasDst <- maybeStopProperty "Dst wallet doesn't exist" =<< lift (WS.getWalletTxHistory dstWalId)
-        void $ expectedOne "TxMeta for dst wallet" txMetasDst
+    -- (!!) is safe here as group will guarantee that
+    let countedWalIds = map (length &&& (!! 0)) . group . sort $ walId:dstWalIds
+    forM_ countedWalIds $ \(c, wId) -> do
+        txMetas <- maybeStopProperty "Wallet doesn't exist" =<< lift (WS.getWalletTxHistory wId)
+        -- FIXME: I am sure that if there was one output, there will
+        -- be exactly one txMeta. If there are multiple outputs tx meta
+        -- count is a bit complicated to calculate (not sure yet)
+        when (c == 1) $
```

Comment:
It should be one `TxMeta` for each mentioned wallet. It's very strange if there is greater than one. I would investigate why this statement is needed to pass tests. A valid number of `TxMeta`s guarantees correct tx history.

Reply:
> It should be one TxMeta for each mentioned wallet

right, I got it confused with `CTxMeta` . I will fix that

Reply:
Ok I am wrong, in a real system, it should be one for all mentioned destination addresses.
It should be one only for source addresses in the test and zero for others addresses

Reply:
I suppose only for source address `c == 1` and it works

Reply:
yes, I agree.